### PR TITLE
PRJ-767: Plugin secure connection support

### PR DESCRIPTION
**Pull request by IKrukov-HORIS**
- HTTPs connection support for auto-generated or user-imported certificates
- plugin settings page added

### CODE REVIEW

**Comment by SerVB** [comment]
Please request my review when it's not a draft

**Review by SerVB — CHANGES_REQUESTED** [review]
Thank you very much, I've checked, it works. Please resolve warnings from Qodana for the changed lines.

Also, I've collected some problems that are mostly UI/UX related:

image

1. When opening the More Settings window, options aren't synchronized with the previous window. When exiting from More Settings, options are passed to the previous window (both if we click OK or Cancel). To reproduce: generate a random password in the first window, open More Settings (newly generated password isn't applied), close More Settings (newly generated password is reset).
2. "Use names in URLs" – this checkbox affects only the shown invitation links, right? Then let's move it to that section to make it clearer.
3. Can we maybe reuse the Connections panel in both windows? For now, it's reused too, but it looks inconsistent. Also, we don't have this beautiful panel header line in the first panel, maybe we can add it there. But the header text itself is clearer in the first panel, "Listen on".
4. Let's move "start automatically..." and "require confirmation..." checkboxes to the bottom (in a new section like "Misc") – this will make the "Listen on" settings group the first group in both windows which can be more intuitive for a user.
5. (probably it was even before this PR, but we can resolve it now) Links (RW and RO) aren't aligned. We can align them, for example, on the right, like we align textfields for passwords.
6. Let's use the same header line for "Invitation links" too.
7. It's cool that we can enable and disable HTTPS when Projector is stopped and then start it with the changed protocol. However, to make this "Use HTTPS" checkbox active after stopping, I have to reopen the config window.

File: projector-plugin/CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,9 @@ Notable changes to this project are documented in this file.
 
 # 1.9.0
 
+## Added
+- HTTPs support
+
```
[the remark this answers is not in the document]
Comment (IKrukov-HORIS) [reply projector-plugin/CHANGELOG.md:9]: Fixed

File: projector-plugin/src/main/kotlin/org/jetbrains/projector/plugin/ui/SessionDialog.kt [hunk 2]
```diff
@@ -25,64 +25,72 @@
...
-  val confirmConnection: Boolean get() = requireConnectConfirmation.isSelected
-  val autostart: Boolean get() = autostartProjector.isSelected
-  private val urlAddress: String get() = urlHostsList.selected?.address ?: ""
+  private val urlAddress: String
+    get() {
+      if (useNamesInURLs.isSelected) {
+        return urlHostsList.selected?.name ?: ""
+      }
+
+      return urlHostsList.selected?.address ?: ""
```
[the remark this answers is not in the document]
Comment (IKrukov-HORIS) [reply projector-plugin/src/main/kotlin/org/jetbrains/projector/plugin/ui/SessionDialog.kt]: Thank you for catching this, fixed.

File: projector-plugin/src/main/kotlin/org/jetbrains/projector/plugin/Utils.kt [hunk 3]
```diff
@@ -161,17 +166,27 @@ private const val SUBSYSTEM = "PROJECTOR_SERVICE_CONFIG"
...
 
+const val RANDOM_PASSWORD_LEN = 11
+
+fun generatePassword(): String {
+  val charPool: List<Char> = ('a'..'z') + ('A'..'Z') + ('0'..'9')
+  return (1..RANDOM_PASSWORD_LEN)
+    .map { Random.nextInt(0, charPool.size) }
+    .map(charPool::get)
+    .joinToString("")
+}
```
Comment (SerVB) [inline projector-plugin/src/main/kotlin/org/jetbrains/projector/plugin/Utils.kt:188]: Was before this PR, but can be improved now:

```suggestion
fun generatePassword(): String {
  val charPool = ('a'..'z') + ('A'..'Z') + ('0'..'9')
  return generateSequence { charPool.random() }
    .take(RANDOM_PASSWORD_LEN)
    .joinToString("")
}
```

File: projector-plugin/src/main/kotlin/org/jetbrains/projector/plugin/ProjectorService.kt [hunk 4]
```diff
@@ -272,5 +295,6 @@ class ProjectorService : PersistentStateComponent<ProjectorConfig> {
 
   override fun loadState(state: ProjectorConfig) {
     config = state
+    setSystemProperty(ProjectorServer.ENABLE_CONNECTION_CONFIRMATION, if (config.confirmConnection) "true" else "false")
```
Comment (SerVB) [inline projector-plugin/src/main/kotlin/org/jetbrains/projector/plugin/ProjectorService.kt:298]: Was before this PR but can be improved now (here and in other places):

```suggestion
    setSystemProperty(ProjectorServer.ENABLE_CONNECTION_CONFIRMATION, config.confirmConnection.toString())
```

**Review by SerVB — APPROVED** [review]
Only minor comments, feel free to merge

**Pull request merged**